ros2bag_tools/ros2bag_tools/exporter/image.py

```diff
@@ -33,6 +34,97 @@ def conversion_code(bayer_encoding, desired_encoding, demosaicing):
         colorspace = 'BGR'
     return getattr(cv, f'COLOR_BAYER_{pattern}2{colorspace}{algorithm}')
 
+class ImageMsgWriter:
+    def __init__(self):
+        self.image_bridge = CvBridge()
+    @staticmethod
+    def get_message_type():
+        """Get the supported message type.
+        """
+        return Image
+    @staticmethod
+    def convert_encoding(img, img_encoding, encoding, demosaicing):
+        """Convert colorspace encoding and debayer if necessary.
+        """
+        if img_encoding.startswith('bayer_') and demosaicing:
+            # use custom debayering
+            # cvtColor2 unfortunately doesn't expose a demosaicing algorithm parameter
+            conv_code = conversion_code(
+                img_encoding, encoding, demosaicing)
+            img = cv.cvtColor(img, conv_code)
+        else:
+            try:
+                img = cvtColor2(img, img_encoding, encoding)
+            except RuntimeError as e:
+                raise CvBridgeError(e)
+        return img
+
+    def process(self, img_msg, encoding, demosaicing, ext, input_encoding=None, params=None):
+        """Write encoded image to buffer and return if successfull True, buf otherwise False, buf.
+        """
+        img = self.image_bridge.imgmsg_to_cv2(msg)
+        if encoding != 'passthrough':
+            in_encoding = img_msg.encoding if not input_encoding else input_encoding
+            img = ImageMsgWriter.convert_encoding(img, in_encoding, encoding, demosaicing)
+        reval, buf = cv.imencode(ext, img, params)
+        return buf
+
+class CompressedImageMsgWriter:
+    def __init__(self):
+        self.image_bridge = CvBridge()
+
+    @staticmethod
+    def get_message_type():
+        """Get the supported message type.
+        """
+        return CompressedImage
+    @staticmethod
+    def get_format(ext):
+        """Get compressed format from extension.
+        """
+        ext = ext.lower()
+        if ext == '.jpg' or ext == '.jpeg':
+            form = 'jpeg'
+        else:
+            form = ext
+        return form
+        
+    @staticmethod
+    def cv2_to_enc(cvim):
+        """Guess default encoding from shape.
+        """
+        if len(cvim.shape) < 3:
+            encoding = 'mono'
+        elif cvim.shape[2] == 3:
+            encoding = 'bgr'
+        elif cvim.shape[2] == 4:
+            encoding = 'bgra'
+        return f'{encoding}{cvim.itemsize*8}'
+
+    def process(self, img_msg, encoding, demosaicing, ext, input_encoding=None, params=None):
+        """Write encoded image to buffer and return if successfull True, buf otherwise False, buf.
+        """
+        img = None
+        convert = encoding != 'passthrough'
+        if convert or img_msg.format != CompressedImageMsgWriter.get_format(ext):
+            img = self.image_bridge.compressed_imgmsg_to_cv2(img_msg)
+        if convert:
+            in_encoding = input_encoding if input_encoding else CompressedImageMsgWriter.cv2_to_enc(img)
```

Comment:
IMO, the encoding should be

1. assumed to be the input encoding if the argument is passed by the user -> done here
2. read from the format field of the compressed msg -> missing?
3. guessed from the shape of the data array -> done here

Reply:
Thanks, I didn't know about this format field structure. You're right, we should consider the stored encoding